src/util/sighash.rs

```diff
@@ -0,0 +1,912 @@
+    pub fn new(script: &'s Script, code_separator_pos: u32, leaf_version: u8) -> Self {
+        ScriptPath {
+            script,
+            code_separator_pos,
+            leaf_version,
+        }
+    }
+    /// Create a new ScriptPath structure using default values for `code_separator_pos` and `leaf_version`
+    pub fn with_defaults(script: &'s Script) -> Self {
+        Self::new(script, 0xFFFFFFFFu32, LEAF_VERSION_TAPSCRIPT)
+    }
+}
+
+impl From<LegacySigHashType> for SigHashType {
+    fn from(s: LegacySigHashType) -> Self {
+        match s {
+            LegacySigHashType::All => SigHashType::All,
+            LegacySigHashType::None => SigHashType::None,
+            LegacySigHashType::Single => SigHashType::Single,
+            LegacySigHashType::AllPlusAnyoneCanPay => SigHashType::AllPlusAnyoneCanPay,
+            LegacySigHashType::NonePlusAnyoneCanPay => SigHashType::NonePlusAnyoneCanPay,
+            LegacySigHashType::SinglePlusAnyoneCanPay => SigHashType::SinglePlusAnyoneCanPay,
+        }
+    }
+}
+
+impl SigHashType {
+    /// Break the sighash flag into the "real" sighash flag and the ANYONECANPAY boolean
+    pub(crate) fn split_anyonecanpay_flag(self) -> (SigHashType, bool) {
+        match self {
+            SigHashType::Default => (SigHashType::Default, false),
+            SigHashType::All => (SigHashType::All, false),
+            SigHashType::None => (SigHashType::None, false),
+            SigHashType::Single => (SigHashType::Single, false),
+            SigHashType::AllPlusAnyoneCanPay => (SigHashType::All, true),
+            SigHashType::NonePlusAnyoneCanPay => (SigHashType::None, true),
+            SigHashType::SinglePlusAnyoneCanPay => (SigHashType::Single, true),
+            SigHashType::Reserved => (SigHashType::Reserved, false),
+        }
+    }
+}
+
+impl<R: Deref<Target = Transaction>> SigHashCache<R> {
+    /// Compute the sighash components from an unsigned transaction and auxiliary
+    /// in a lazy manner when required.
+    /// For the generated sighashes to be valid, no fields in the transaction may change except for
+    /// script_sig and witnesses.
+    pub fn new(tx: R) -> Self {
+        SigHashCache {
+            tx,
+            common_cache: None,
+            taproot_cache: None,
+            segwit_cache: None,
+        }
+    }
+
+    /// Encode the BIP341 signing data for any flag type into a given object implementing a
+    /// std::io::Write trait.
+    pub fn taproot_encode_signing_data_to<Write: io::Write>(
+        &mut self,
+        mut writer: Write,
+        input_index: usize,
+        prevouts: &Prevouts,
+        annex: Option<Annex>,
+        script_path: Option<ScriptPath>,
+        sighash_type: SigHashType,
+    ) -> Result<(), Error> {
+        prevouts.check_all(&self.tx)?;
+
+        let (sighash, anyone_can_pay) = sighash_type.split_anyonecanpay_flag();
+
+        // epoch
+        0u8.consensus_encode(&mut writer).unwrap();
```

Comment:
This is assuming the `writer` is an `io::Write` that doesn't error, right? I suppose that should at least be documented, but ideally we'd propagate that error as an `Error::Writer(io::Error)` variant.

Reply:
I realize that we use this method internally with a safe Write, but since it's a public method, someone else might not..

Reply:
There was a variant, but there was a suggestion to remove it https://github.com/rust-bitcoin/rust-bitcoin/pull/628#discussion_r671166566 since probably never exercised. Now I see calling `*_encode_signing*` with custom Writer may throw an Error, @sgeisler I think I need to remove 6d4db61 and change the variant comment, do you agree?

Reply:
Right, I totally missed these methods were public. So either remove d4db61 or make them private I guess (was is their intended public use case?).

Reply:
Good point that I think they might be really hard to use for anything else than hashing. The signing data is not very structured, so I doubt anyone wants to write it into anything else than a hasher. Maybe someone would want to use them with a hardware-accelerated sha256 engine or so?

Reply:
external writers are used for accessing serialized data, see https://github.com/rust-bitcoin/rust-bitcoin/issues/483

I removed 6d4db61 and added 1da9706db98b7e873b1209052cdb9648567dab7a

I wrapped `ErrorKind` instead of `io::Error` because I think it's a good tradeoff allowing to auto derive some useful traits (Copy, Eq...)

Reply:
ah there you go

Reply:
and yeah ack, io::Error is an annoying one.

Reply:
So this will probably require changes to #603 or the other way around depending on merge order.